mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java

```diff
@@ -52,103 +41,147 @@
+                .join()
+                .orElseThrow(() -> new MailboxNotFoundException(path));
+        } catch (CompletionException e) {
+            if (e.getCause() instanceof InvalidQueryException) {
+                if (StringUtils.containsIgnoreCase(e.getCause().getMessage(), VALUES_MAY_NOT_BE_LARGER_THAN_64_K)) {
+                    throw new TooLongMailboxNameException("too long mailbox name");
+                }
+                throw new MailboxException("It has error with cassandra storage", e.getCause());
             }
-            throw new MailboxException("It has error with cassandra storage", e);
+            throw e;
         }
     }
 
     @Override
     public Mailbox findMailboxById(MailboxId id) throws MailboxException {
         CassandraId mailboxId = (CassandraId) id;
-        ResultSet resultSet = session.execute(select(FIELDS).from(TABLE_NAME).where(eq(ID, mailboxId.asUuid())));
-        if (resultSet.isExhausted()) {
-            throw new MailboxNotFoundException(id.serialize());
-        } else {
-            return mailbox(resultSet.one());
-        }
+        return mailboxDAO.retrieveMailbox(mailboxId)
+            .join()
+            .orElseThrow(() -> new MailboxNotFoundException(id.serialize()));
     }
 
     @Override
     public List<Mailbox> findMailboxWithPathLike(MailboxPath path) throws MailboxException {
         Pattern regex = Pattern.compile(constructEscapedRegexForMailboxNameMatching(path));
-        return getMailboxFilteredByNamespaceAndUserStream(path.getNamespace(), path.getUser())
-            .filter((row) -> regex.matcher(row.getString(NAME)).matches())
-            .map(this::mailbox)
-            .collect(Collectors.toList());
+        return mailboxPathDAO.listUserMailboxes(path.getNamespace(), path.getUser())
+            .thenApply(stream -> stream.filter(idAndPath -> regex.matcher(idAndPath.getMailboxPath().getName()).matches()))
+            .thenApply(stream -> stream.map(CassandraMailboxPathDAO.CassandraIdAndPath::getCassandraId))
+            .thenApply(stream -> stream.map(mailboxDAO::retrieveMailbox))
+            .thenCompose(CompletableFutureUtil::allOf)
+            .thenApply(stream -> stream
+                .flatMap(OptionalConverter::toStream)
+                .collect(Guavate.<Mailbox>toImmutableList()))
+            .join();
     }
 
     @Override
     public void save(Mailbox mailbox) throws MailboxException {
         Preconditions.checkArgument(mailbox instanceof SimpleMailbox);
         SimpleMailbox cassandraMailbox = (SimpleMailbox) mailbox;
+
+        CassandraId cassandraId = retrieveId(cassandraMailbox);
+        cassandraMailbox.setMailboxId(cassandraId);
+        if (mailbox.getName().length() >= MAX_SIZE_FOR_MAILBOX_NAME) {
+            throw new TooLongMailboxNameException("too long mailbox name");
+        }
+        try {
+            boolean applied = mailboxDAO.retrieveMailbox(cassandraId)
```

Comment:
1. Read mailboxDAO to retrieve potential old path
    2. (thenCompose) Inset new Path in mailboxPathDAO
    3. thenCompose if success, parallel(delete old path with mailboxPathDAO, save in mailboxDAO)

Add a test on moving mailbox. If new name too long, old mailboxPath should still exists